rfcs/0006-merkle-tree/0006-merkle-tree.md

```diff
@@ -0,0 +1,145 @@
+---
+Number: 0006
+Category: Standards Track
+Status: Draft
+Author: Ke Wang
+Organization: Nervos Foundation
+Created: 2018-12-01
+---
+
+# Complete Binary Merkle Tree
+
+## Introduction
+
+CKB uses Complete Binary Merkle Tree(CBMT) to generate Merkle Root and Merkle Proof for a static list of items, such as Transactions Root. Basically, CBMT is a complete binary tree as well as a full tree, it has the same node order with the complete binary tree. Compare with other Merkle trees, the hash computation of CBMT is minimal, as well as the proof size.
+
+## Tree Struct
+
+Tree with 6 items(the hashes are [T0, T1, T2, T3, T4, T5]) and tree with 7 items(the hashes are [T0, T1, T2, T3, T4, T5, T6]) is shown below:
+
+```
+        with 6 items                       with 7 items
+
+              B0                                 B0
+             /  \                               /  \
+           /      \                           /      \
+         /          \                       /          \
+       /              \                   /              \
+      B1              B2                 B1              B2
+     /  \            /  \               /  \            /  \
+    /    \          /    \             /    \          /    \
+   /      \        /      \           /      \        /      \  
+  B3      B4      TO      T1         B3      B4      B5      T0
+ /  \    /  \                       /  \    /  \    /  \
+T2  T3  T4  T5                     T1  T2  T3  T4  T5  T6
+```
+
+CBMT can be stored in an array, so the two trees above can be stored as:
+
+```
+[B0, B1, B2, B3, B4, T0, T1, T2, T3, T4, T5]
+[B0, B1, B2, B3, B4, B5, T0, T1, T2, T3, T4, T5, T6]
+```
+
+Suppose we have n items, the size of array would be 2n-1, the index of item i(start at 0) is i+n-1. For node at i, the index of its parent is (i-1)/2, the index of its sibling is (i+1)^1-1 and the indexes of its children are [2i+1, 2i+2].
+
+Specially, the tree with 0 item is empty(0 node) and its root is H256::zero.
+
+## Merkle Proof
+
+Merkle Proof can provide a proof for existence of one or more items. Only sibling of the nodes along the path that form leaves to root, excluding the nodes already in the path, should be included in the proof. Moreover, the nodes should presented in the index's descending order. If we want to show that [T1, T4] is in the list of 6 items above, only nodes [T5, T0, B3] should be included in the proof.
```

Comment:
> the nodes should presented in the index's descending order

why?

Reply:
It would be helpful to explain why the merkle proof is very succinct in CBMT.

Reply:
simple rule with a simple algorithm